### PR TITLE
fix: blocking behaviour under load

### PR DESCRIPTION
**Pull request by trentm**
Update to http client which fixes/improves comms with apm-server.

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings
- [x] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by trentm** [comment]
That failing integration test https://apm-ci.elastic.co/blue/organizations/jenkins/apm-integration-tests-selector-mbp/detail/master/15669/pipeline

```
...
[2021-04-05T22:30:07.129Z] =========================== short test summary info ============================
[2021-04-05T22:30:07.129Z] FAILED tests/agent/test_nodejs.py::test_request_express - AssertionError: Exp...
[2021-04-05T22:30:07.129Z] FAILED tests/agent/test_nodejs.py::test_express_error - AssertionError: Expec...
[2021-04-05T22:30:07.129Z] FAILED tests/agent/test_nodejs.py::test_conc_req_express - AssertionError: qu...
[2021-04-05T22:30:07.129Z] FAILED tests/agent/test_nodejs.py::test_conc_req_node_foobar - AssertionError...
[2021-04-05T22:30:07.129Z] =========== 4 failed, 11281 warnings, 12 rerun in 885.55s (0:14:45) ============
[2021-04-05T22:30:07.129Z] make: *** [Makefile:101: test-agent-nodejs] Error 1
[2021-04-05T22:30:07.707Z] Makefile:134: recipe for target 'dockerized-test' failed
[2021-04-05T22:30:07.707Z] make[1]: *** [dockerized-test] Error 2
[2021-04-05T22:30:07.707Z] make[1]: Leaving directory '/var/lib/jenkins/workspace/ration-tests-selector-mbp_master/src/github.com/elastic/apm-integration-testing'
[2021-04-05T22:30:07.707Z] Makefile:131: recipe for target 'docker-test-agent-nodejs' failed
[2021-04-05T22:30:07.707Z] make: *** [docker-test-agent-nodejs] Error 2
script returned exit code 2
```

worked for me when trying on my dev machine: https://gist.github.com/trentm/cc20ef0cc4a5d62aa9d8a95fb661df4c

I'm not sure what that failure is. I'll try running it again.

**Comment by trentm** [comment]
jenkins run the tests

**Comment by trentm** [comment]
I can repro the integration tests failures via:

```
% cd .../apm-integration-testing
% export BUILD_OPTS="--nodejs-agent-package elastic/apm-agent-nodejs#e7c50f21fa0bf3f709b5ca047781222c411d754d --opbeans-node-agent-branch e7c50f21fa0bf3f709b5ca047781222c411d754d --build-parallel"
% export ELASTIC_STACK_VERSION=8.0.0
% .ci/scripts/agent.sh nodejs nodejs-express
```

**Review by astorm — CHANGES_REQUESTED** [review]
A reminder that we'll want to undo https://github.com/elastic/apm-agent-nodejs/pull/2033 once this one lands.

**Comment by trentm** [comment]
jenkins run the tests

**Comment by trentm** [comment]
I can repro the current Integration Tests failure locally via:

```
python3 scripts/compose.py start 8.0.0 --nodejs-agent-package elastic/apm-agent-nodejs#73cf4d3dd302cc9ec10945be59b5b1d859642081 --opbeans-node-agent-branch 73cf4d3dd302cc9ec10945be59b5b1d859642081 --build-parallel   --with-agent-nodejs-express   --with-opbeans-node   --no-apm-server-dashboards   --no-apm-server-self-instrument   --apm-server-agent-config-poll=1s   --force-build --no-xpack-secure
```

FWIW.

**Comment by trentm** [comment]
jenkins run the tests

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
     "coverage:report": "nyc report --reporter=lcov",
     "test": "./test/script/run_tests.sh",
     "test:cli": "node test/script/cli.js",
-    "test:deps": "dependency-check *.js 'lib/**/*.js' 'test/**/*.js' --no-dev -i async_hooks -i perf_hooks -i parseurl",
+    "test:deps": "dependency-check start.js index.js 'lib/**/*.js' 'test/**/*.js' --no-dev -i async_hooks -i perf_hooks -i parseurl",
     "test:tav": "tav --quiet",
```
Comment (trentm) [inline package.json:19]: Reviewer note: I like to have "foo.js" and "play.js" files (etc.) in my working copy. It is unhelpful when those break `npm test` because they use some dev/test dependency, e.g.:

```
% npx dependency-check *.js 'lib/**/*.js' 'test/**/*.js' --no-dev -i async_hooks -i perf_hooks -i parseurl
Fail! Dependencies not listed in package.json: blocked-at
```

where "blocked-at" is only being used here for dev/testing (it was installed via `npm install --no-save blocked-at`.


I could move to a separate PR if you prefer.

File: test/_apm_server.js [hunk 2]
```diff
@@ -56,15 +56,12 @@ function APMServer (agentOpts, mockOpts = { expect: [] }) {
       assert.strictEqual(req.method, 'POST', `Unexpected HTTP method: ${req.method}`)
       assert.strictEqual(req.url, '/intake/v2/events', `Unexpected HTTP url: ${req.url}`)
 
-      self.on('close', () => {
-        res.end()
-      })
-
```
Comment (trentm) [inline test/_apm_server.js:62]: Reviewer note: we shouldn't only end the response from the APM server when we server.close(). Below we properly respond.

File: test/agent.js [hunk 3]
```diff
@@ -789,31 +792,36 @@ test('#flush()', function (t) {
...
           t.pass('should call flush callback after flushing the queue')
+          t.end()
         })
       })
       .on('data-transaction', function (data) {
-        t.strictEqual(data.name, 'foo')
-        t.end()
+        // 1. The APM server should receive the transaction, and then ...
+        t.strictEqual(data.name, 'foo', 'APM server received the "foo" transaction')
       })
```
Comment (trentm) [inline test/agent.js:804]: Reviewer note: All of the test fixes are of this form. Now that the http client actually waits for a response from the APM server before concluding, the callback to `.flush(cb)` or to `.captureError(cb)` is reliably called *after* the APM server receives the data (i.e. the "data-transaction" event from the mock APM server here).

**Review by trentm — COMMENTED** [review]
Reviewer notes.

File: test/_apm_server.js [hunk 4]
```diff
@@ -78,6 +75,10 @@ function APMServer (agentOpts, mockOpts = { expect: [] }) {
 
         index++
       })
+      parsedStream.on('end', function () {
```
Comment (astorm) [inline test/_apm_server.js:78]: Question: Why end with a 202? Is this just mimicking APM server's HTTP response code, or is there more going on here?

**Review by astorm — APPROVED** [review]
New docs and changelog look good/accurate, changes to the tests look required/reasonable. 

I just have one non-question about the `202` status code in `_apm_server.js` -- but it's non-blocking.  Approving.

File: test/_apm_server.js (hunk 4 above)
Comment (trentm) [reply test/_apm_server.js:78]: Yes, just mimicking APM server's HTTP response code.

Some background: In earlier work of the `makeInjestRequest()` re-write in the http client I had initially only accepted a "202" response from APM server. However, I ended up relaxing that to accepting any "2xx" response because we have test suites using mock APM servers (like this one) that just `res.end()` defaulting to a "200" response.
